src/userSurvey.js

```diff
@@ -184,43 +291,118 @@ export async function setSurvey(options) {
+            setPopupToNoPromptPage();
+            browser.webRequest.onBeforeRequest.removeListener(surveyCompletionUrlListener);
+        }, { type: "webScience.userSurvey.cancelSurvey" });
+        messaging.onMessage.addListener(() => {
+            openSurveyInNewTab();
+        }, { type: "webScience.userSurvey.openSurvey" });
+    }
+
+    listenersRegistered = true;
 }
 
 /**
  * Each study participant has a persistent survey ID, generated with
  * the id module. The ID is automatically added as a parameter to
  * the survey URL, enabling researchers to import survey data from an
- * external platform and sync it with Rally data.
+ * external platform and sync it with Rally data. This method returns the
+ * survey ID, generating it if it does not already exist.
  * @returns {string} - The participant's survey ID.
  */
 export async function getSurveyId() {
-    return await storageSpace.get("surveyId");
+    initializeStorage();
+    let surveyId = await storageSpace.get("surveyId");
+    if (surveyId === null) {
+        surveyId = id.generateId();
+        await storageSpace.set("surveyId", surveyId);
+    }
+    return surveyId;
+}
+
+/**
+ * Gets the status of the current survey. Can be used if a
+ * subsequent survey depends on the status of the previous survey.
+ * @returns {string|null} - The status of the survey (either "completed",
+ * "cancelled", or "active") or null if there is no survey.
+ */
+export async function getSurveyStatus() {
+    initializeStorage();
+
+    const surveyDetails = await storageSpace.get("surveyDetails");
+    const surveyCompleted = await storageSpace.get("surveyCompleted");
+    const surveyCancelled = await storageSpace.get("surveyCancelled");
+
+    if (!surveyDetails) {
+        return null;
+    } else if(surveyCompleted) {
+        return "completed";
+    } else if(surveyCancelled) {
+        return "cancelled";
+    } else {
+        return "active";
+    }
 }
+
+/**
+ * Gets the name of the current survey.
+ * @returns {string|null} - The name of the current survey. Returns null
+ * if there is no current survey.
+ */
+export async function getSurveyName() {
+    initializeStorage();
+    const surveyDetails = await storageSpace.get("surveyDetails");
+    return surveyDetails ? surveyDetails.surveyName : null;
+}
+
+/**
+ * End the current survey. Should be called before a subsequent survey is started.
+ */
+export async function endSurvey() {
+    // Stop prompting for the survey.
+    setPopupToNoPromptPage();
+
+    // If there is an existing survey reminder timeout, clears the timeout.
+    clearTimeout(reminderTimeoutId);
+
+    // Remove any previously added listener for browser.webRequest.onBeforeRequest
+    // that checks for the survey completion URL.
+    browser.webRequest.onBeforeRequest.removeListener(surveyCompletionUrlListener);
+
+    initializeStorage();
+
+    // Clears the the data in storage for the current survey.
+    await storageSpace.set("lastSurveyRequest", null);
```

Comment:
Hm, just wondering why `[clear()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage/StorageArea/clear)` is not exposed from the storage module?

Reply:
We don't currently store all the keys associated with a `KeyValueStorage` storage space (which is just a key namespace subset of extension storage), so there isn't a way to clear the storage space without iterating all the keys in extension storage. We could add a `clear` function (and a `remove` function), those just haven't been necessary yet.